### ISSUE TITLE
Azure cluster prepare

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#908]
Implement neonKUBE setup for Azure:

- [ ] Provision
     - [x] Support the new UltraSSD storage type and the new drive sizes for the older storage types
     - [x] Refactor/cleanup Azure related cluster definition and node options
          - [x] Support default node instance types/sizes
          - [x] Support default drive types and sizes
          - [x] Drive size properties should be converted from int GiB to string ByteUnits
          - [x] Rename Azure related cluster definition properties for ones that don't make a lot of sense.
          - [x] Instance type/sizes should be specified by strings and not an enum.  Azure is adding more instance types than we can really keep up with.  We also need to stop attempting to keep track of instance size details like RAM, cores, etc. ourselves.
&nbsp;
     - [x] `AzureHostingManager` improved verification:
          - [x] Ensure that the Azure environment is available to the account
          - [x] Ensure that the Azure region is available to the account
          - [x] Ensure that all cluster VM sizes are available in the region
          - [x] Ensure that UltraSSD is actually available for in the region and the cluster VM sizes
          - [x] Ensure that the cluster VM sizes have the minimal required RAM and CPU cores
          - [x] Verify node role is valid
          - [x] Ensure that explicit node addresses are within the node subnet and don't conflict with any reserved addresses.
&nbsp;
     - [x] Mind the Gap
          - [x] Node DNS currently default to Google servers (8.8.8.8 & 8.8.4.4) for all clouds.  This should default to the cloud's DNS.  The may require a change in the node setup scripts.
          - [x] Automatically assign VNET addresses to nodes that don't already have them.  Try to place master nodes towards the beginning of the subnet and avoid reserved addresses.
&nbsp;
     - [x] Rename **"Azure Drives"** to **"Azure Disks"** to be consistent with the Azure lingo.
     - [x] Create resource group if necessary
     - [x] Create proximity placement group if requested
     - [x] Create the VM availability sets
     - [x] Create the VNET and its subnet
     - [x] Create and manage the subnet NSG and its rules
          - [x] Ingress rules with source IP white/black listing
          - [x] Temporary management SSH NAT rules with source IP white/black listing
     - [x] Create public IP address
     - [x] Create NICs
     - [x] Create the VMs and configure their disks
     - [x] Create the external load balancer
     - [x] Add code to manage the load balancer:
          - [x] Ingress rules
          - [x] Temporary management SSH NAT rules
&nbsp;
     - [x] The Canonical Ubuntu 20.04 images don't appear to include the Azure Agent:
          https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/agent-linux#installation

          **JEFF UPDATE:** Never mind, the agent is running, it just takes a while for that to be reported on the Azure portal.
&nbsp;
     - [x] Ubuntu 20.04 has different Azure images for Gen1 and Gen2 capable VM sizes.  We need to use the Gen2 image for VMs that support it.  Here's the command to list the available Ubuntu 20.04 images:
          ```
          az vm image list --publisher Canonical --all > c:/temp/images.txt
          ```
          Here's some info on using images: https://docs.microsoft.com/en-us/cli/azure/vm/image?view=azure-cli-latest
          More info on accepting an image offer: https://docs.microsoft.com/en-us/cli/azure/vm/image/terms?view=azure-cli-latest
&nbsp;
- [x] Prepare
     - [x] This is almost working.  The problem is that the mounted Azure disk device may change after reboots.  I've seen it as `/dev/sda` and also `dev/sdc`.  The link below discusses this:
          https://docs.microsoft.com/en-us/azure/virtual-machines/troubleshooting/troubleshoot-device-names-problems

**Comment by jefflill** [comment]
Let's start by how we'll think about Azure before writing a bit of a specification.  Here are some good links to get started with:

https://docs.microsoft.com/en-us/azure/availability-zones/az-overview
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/manage-availability
https://docs.microsoft.com/en-us/azure/virtual-network/virtual-networks-overview
https://docs.microsoft.com/en-us/azure/virtual-machines/linux/scheduled-events
https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview

### Availability

https://docs.microsoft.com/en-us/azure/virtual-machines/windows/manage-availability

Azure provides mechanisms for deploying multiple VMs such that the chance that all of them becoming unavailable at the same time will (hopefully) be rare.  VM availability can be impacted by local hardware failures in the physical rack hosting the VM, a failure of an Azure availability zone, a hardware failure of the machine hosting the VM, or a software update required by the hosting machine.

Production neonKUBE clusters will generally deploy 3 Kubernetes manager nodes.  These nodes coordinate to form a highly reliable source of truth for the expected state of of the services and pods running on the cluster.  Managers rely on a RAFT style leader election mechanism combined with mechanisms for reliably replicating state across the managers.  RAFT requires a majority of the managers (2 out of 3 in this case) to be available for the cluster to operate properly.  This means that cluster deployment must take care to ensure that only one manager at a time is at risk of failing or being updated and rebooted.

Most neonKUBE operators will also need this sort of resilience.  For example, they may wish to deploy multiple web servers such that some will always be available in the face of failures or the may also be deploying a multi-node database cluster such that data is replicated to prevent a single source of failure.

Azure has several related concepts:

* **Region:** A collection of datacenters sited close together with dedicated low-latency networking.  Each datacenter is equipped with isolated power, cooling, and networking.

* **Availablity Zone:** A unique set of one or more datacenters in a region where customer resources can be hosted such that they can be isolated from failures of other availability zones in the region.

**Comment by jefflill** [comment]
[placement policy](https://cloud.google.com/compute/docs/instances/define-instance-placement)

**Comment by jefflill** [comment]
Here's another link discussing the MSFT Partner Program.  We'd need to do this to publish VM images.

https://docs.microsoft.com/en-us/azure/marketplace/marketplace-faq-publisher-guide

**Comment by jefflill** [comment]
For neonHIVE, I created a long list of Azure VM types describing their CPUs, RAM, and NIC/drive capacities (which I'm sure is very out of date by now).  One of the goals for cluster setup is that we want to detect any problems with the cluster configuration early in the setup process; we don't want the customer to go well into cluster setup only to find that a VM type doesn't have enough RAM or something.

One problem with this list of instance types was that it assumed that all instance types are available in all Azure datacenters and this isn't true.  It's very common for MSFT to roll out new instance types to DCs over time and perhaps retire them in the same way.

I took a look at how OpenShift does this on Azure:

https://azure.microsoft.com/en-us/pricing/details/openshift/

This is **very interesting**.  OpenShift has a list of instance types operators can select from.  This includes just one rather beefy type for manager nodes: D8sv3 8 core 32GiB RAM at $0.44/hr or $316/mo.  OpenShift **does not** add additional fees for manager nodes (like we're thinking).  The crazy thing is that they claim a worker node this big can still handle only 100 worker nodes in the cluster.  We're going to have to do some capacity testing to understand these limits better.

For worker nodes, OpenShift charges by hour on a sliding scale based on the capability of the instance type:

![image](https://user-images.githubusercontent.com/4908431/88485850-271eda00-cf2e-11ea-8bde-f44654328600.png)

The costs are pretty outrageous: 

![image](https://user-images.githubusercontent.com/4908431/88486215-04da8b80-cf31-11ea-84e7-0cbb1e0cc52b.png)
![image](https://user-images.githubusercontent.com/4908431/88486243-3ce1ce80-cf31-11ea-8737-5b462d091ba4.png)

So the smallest reasonable cluster will cost $5.4K/mo or $3.8K/mo with reserved instances.  

![image](https://user-images.githubusercontent.com/4908431/88486384-18d2bd00-cf32-11ea-95f3-1963de34c352.png)

That really leaves a lot of pricing room for neonCLOUD (BTW: I already own **neoncloud.io**) for just this purpose.

**Comment by jefflill** [comment]
Implemented for v2.2.0
